assets/my-first-radanalytics-app/sparkpi-scala-scalatra.adoc

```diff
@@ -0,0 +1,329 @@
+    }
+}
+```
+
+2) Setup Scalatra initialization and plumbing. Scalatra will want to default this class to being called **ScalatraBootstrap** in the default package. This behavior will be overridden in the bootstrap of the application to encourage better code organization.
+```
+package io.radanalytics.examples.scalatra.sparkpi
+
+import javax.servlet.ServletContext
+import org.scalatra.LifeCycle
+
+class ScalatraInit extends LifeCycle {
+
+    override def init( context : ServletContext ) {
+        context.mount( classOf[ SparkPiServlet ], "/*" )
+    }
+
+}
+```
+3) Implement the **Main** class, which will bootstrap Jetty and bind the **SparkPiServlet** so that it can handle requests. Notice the addition of an init parameter, which overrides the Scalatra default mentioned in the previous step.
+
+```scala
+package io.radanalytics.examples.scalatra.sparkpi
+
+import org.eclipse.jetty.server.Server
+import org.eclipse.jetty.servlet.DefaultServlet
+import org.eclipse.jetty.webapp.WebAppContext
+import org.scalatra.servlet.ScalatraListener
+
+object Main {
+
+    def main( args : Array[ String ] ) : Unit = {
+        val port = 8080 //TODO - do I need to make the port configurable/dynamic?
+        val server = new Server( port )
+        val context = new WebAppContext()
+
+        context.setContextPath( "/" )
+        context.setResourceBase( "src/main/webapp" )
+        context.setInitParameter( ScalatraListener.LifeCycleKey, "io.radanalytics.examples.scalatra.sparkpi.ScalatraInit" ) // Override the Scalatra default for ScalatraBootstrap in default package
+        context.addEventListener( new ScalatraListener )
+        context.addServlet( classOf[ DefaultServlet ], "/" ) // handles empty context root
+
+        server.setHandler( context )
+        server.start()
+        server.join()
+    }
+
+}
+```
+
+== Build the Application and Verify Locally
+At this point, the directory structure should look like this:
+
+....
+tutorial-sparkpi-scala-scalatra (root)
+    \_ build.sbt
+    \_ project
+        \_ Dependencies.scala
+        \_ build.properties
+        \_ plugins.sbt
+    \_ src
+        \_ main
+            \_ scala
+                \_ io
+                    \_ radanalytics
+                        \_
+                            \_ examples
+                                \_ scalatra
+                                    \_ sparkpi
+                                        \_ Main.scala
+                                        \_ ScalatraInit.scala
+                                        \_ SparkPi.scala
+                                        \_ SparkPiServlet.scala
+
+    \_ src
+        \_ main
+            \_ scala
+                \_ io
+                    \_ radanalytics
+                        \_
+                            \_ examples
+                                \_ scalatra
+                                    \_ sparkpi
+                                        \_ SparkPiTest.scala
```

Comment:
There are a couple of things going on here. SBT derives a lot of it's functionality by following Maven conventions. Maven has more or less codified what a JVM derived language project should look like in the https://maven.apache.org/guides/introduction/introduction-to-the-standard-directory-layout.html.

The implication here has to do with the build artifacts. Anything placed underneath the `src/main` folder will wind up in the application that is produced at the end of the build. Anything in `src/test` will be compiled and added to the classpath for the testing phase only.

The end result is that the build artifacts do not include any test classes or test resources when they are deployed. This is a standard practice in the JVM community and should be familiar to any Scala/Java/Groovy developer.

Reply:
@elmiko Long story short re: ^^ my previous comment ^^ - this is the most realistic project structure one would have when developing a Scala application. This would also be true for Java applications as well as `src/main` and `src/test` are well codified conventions in organizing project source code.

Reply:
I see what I did there. I'll fix it to make it more clear. Also, good catch on the fact that I just copy and pasted main and didn't change it to test.